lib/airbrake-ruby/truncator.rb

```diff
@@ -1,121 +1,96 @@
+      truncate_object(object, seen << object)
     end
 
-    ##
-    # Reduces maximum allowed size of the truncated object.
-    # @return [Integer] current +max_size+ value
+    # Reduces maximum allowed size of hashes, arrays and strings by half.
+    # @return [void]
     def reduce_max_size
       @max_size /= 2
     end
 
     private
 
-    def truncate(val, seen)
-      case val
-      when String
-        truncate_string(val)
-      when Array, Hash, Set
-        truncate_object(val, seen)
-      when Numeric, TrueClass, FalseClass, Symbol, NilClass
-        val
+    def truncate_object(object, seen)
+      case object
+      when Hash then truncate_hash(object, seen)
+      when Array then truncate_array(object, seen)
+      when Set then truncate_set(object, seen)
+      when String then truncate_string(object)
+      when Numeric, TrueClass, FalseClass, Symbol, NilClass then object
       else
-        stringified_val =
-          begin
-            val.to_json
-          rescue *Notice::JSON_EXCEPTIONS
-            val.to_s
-          end
-        truncate_string(stringified_val)
+        truncate_string(stringify_object(object))
       end
     end
 
     def truncate_string(str)
-      str = replace_invalid_characters!(str)
-      return str if str.length <= @max_size
-      str.slice(0, @max_size) + '[Truncated]'.freeze
+      fixed_str = replace_invalid_characters(str)
+      return fixed_str if fixed_str.length <= @max_size
+      (fixed_str[0...@max_size] + '[Truncated]').freeze
     end
 
-    ##
-    # Replaces invalid characters in string with arbitrary encoding.
-    #
-    # @param [String] str The string to replace characters
-    # @return [String] a UTF-8 encoded string
-    # @note This method mutates +str+ unless it's frozen,
-    #   in which case it creates a duplicate
-    # @see https://github.com/flori/json/commit/3e158410e81f94dbbc3da6b7b35f4f64983aa4e3
-    def replace_invalid_characters!(str)
-      encoding = str.encoding
-      utf8_string = (encoding == Encoding::UTF_8 || encoding == Encoding::ASCII)
-      return str if utf8_string && str.valid_encoding?
-
-      str = str.dup if str.frozen?
-      str.encode!(TEMP_ENCODING, ENCODING_OPTIONS) if utf8_string
-      str.encode!('utf-8', ENCODING_OPTIONS)
+    def stringify_object(object)
+      object.to_json
+    rescue *Notice::JSON_EXCEPTIONS
+      object.to_s
     end
 
     def truncate_hash(hash, seen)
+      truncated_hash = {}
       hash.each_with_index do |(key, val), idx|
-        if idx < @max_size
-          hash[key] = truncate(val, seen)
-        else
-          hash.delete(key)
-        end
+        break if idx + 1 > @max_size
+        truncated_hash[key] = truncate(val, seen)
       end
+
+      truncated_hash.freeze
     end
 
     def truncate_array(array, seen)
-      array.slice(0, @max_size).map! { |val| truncate(val, seen) }
+      array[0...@max_size].map { |elem| truncate(elem, seen) }.freeze
     end
```

Comment:
This is a clean improvement accounting for frozen arrays nicely, but we end up creating two truncated arrays and can refine it to only have one. The `array[0...@max_size]` statement creates a copy of the original array over the requested slice, so that copy isn't frozen. But then we use `map` to avoid modifying the target, which builds another copy of the truncated array. The fix is super simple - we can modify that initially truncated array in-place - your improved code above works fine with `map!`:

```ruby
array[0...@max_size].map! { |elem| truncate(elem, seen) }.freeze
```

For more, see:

```ruby
array = [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 ].freeze
puts "Original: #{ array.object_id }"
short = array[0...6]
puts "Shortened: #{ short.object_id }"
mapped = short.map { | elem | elem }
puts "With 'map': #{ mapped.object_id } / #{ mapped.object_id == short.object_id }"
mapped = short.map! { | elem | elem }
puts "With 'map!': #{ mapped.object_id } / #{ mapped.object_id == short.object_id }"
```

...yields e.g.:

```
Original: 70230550309900
Shortened: 70230550310920
With 'map': 70230550313500 / false
With 'map!': 70230550310920 / true
```

Reply:
Yeah, good catch, I think the only real addition this method needs is the `freeze` call.